### PR TITLE
Rename variable payments to output payment singular references

### PR DESCRIPTION
**Pull request by plisovin**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -33,8 +33,8 @@ Here are the names we are using in the code and specs for the different concepts
...
 >
-> Per participant price £995 >> per participant variable payment £597 (60%) >> per participant variable payment for a retention period £119 (20% of 60%) >> variable payment subtotal for a retention period with 1900 retained participants £226k
+> Per participant price £995 >> per participant output payment £597 (60%) >> per participant output payment for a retention period £119 (20% of 60%) >> output payment subtotal for a retention period with 1900 retained participants £226k
 
 * "Participants" includes both teachers and mentors.
-* Note that the contractual term for "variable payment" is actually "output payment", however we're avoiding this in the code to avoid the inevitable confusion with program input/output terminology. These are payments made based on the performance of the training provider (i.e. their output).
-* "Payment type" for start/retention/completion variable payments.
+* Note that the contractual term for "output payment" is actually "output payment", however we're avoiding this in the code to avoid the inevitable confusion with program input/output terminology. These are payments made based on the performance of the training provider (i.e. their output).
```
Comment (timabell) [inline README.md:39]:
```suggestion
* "Output" payments are payments made based on the performance of the training provider (i.e. their output). Previously known as "variable" in the code.
```

**Comment by timabell** [comment]
To the view the diff with renames detected correctly:

`git show origin/break-up-ecf-feature~ -M40%`

**Review by timabell — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -33,8 +33,8 @@ Here are the names we are using in the code and specs for the different concepts
 > Per participant price £995 >>
 per participant service fee £398 (40%) >> monthly service fee £27k >> total service fee £796k
 >
-> Per participant price £995 >> per participant variable payment £597 (60%) >> per participant variable payment for a retention period £119 (20% of 60%) >> variable payment subtotal for a retention period with 1900 retained participants £226k
+> Per participant price £995 >> per participant output payment £597 (60%) >> per participant output payment for a retention period £119 (20% of 60%) >> output payment subtotal for a retention period with 1900 retained participants £226k
 
 * "Participants" includes both teachers and mentors.
-* Note that the contractual term for "variable payment" is actually "output payment", however we're avoiding this in the code to avoid the inevitable confusion with program input/output terminology. These are payments made based on the performance of the training provider (i.e. their output).
-* "Payment type" for start/retention/completion variable payments.
+* "Output" payments are payments made based on the performance of the training provider (i.e. their output). Previously known as "variable" in the code.
```
Comment (blackrat) [inline README.md:39]:
```suggestion
* "Output payments" are payments made based on the performance of the training provider (i.e. their output). Previously known as "variable payments" in the code.
```

File: lib/services/npq/output_payments.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+      def output_payment_split
+        1.0 / retention_points.length
+      end
+
+      def per_participant_output_payments
+        (per_participant_price * 0.6 * output_payment_split).round(2)
+      end
+
+      def total_output_payments(retained_participants)
```
Comment (blackrat) [inline lib/services/npq/output_payments.rb:28]: singular? Is it a total "Output payment" that seems to be consistent with how it's used in other places.

**Review by blackrat — CHANGES_REQUESTED** [review]

File: lib/services/ecf/payment_calculation.rb [hunk 4]
```diff
@@ -10,7 +10,7 @@ def call
           input: config.to_h,
           output: {
             service_fees: ServiceFees.call(config),
-            variable_payments: VariablePaymentsAggregator.call(config),
+            output_payments: OutputPaymentAggregator.call(config),
```
Comment (blackrat) [inline lib/services/ecf/payment_calculation.rb:13]: This should be singular too. It's still a payment to a provider, even though there are per participant values as part of the breakdown and a "payment over time" element.

File: lib/services/npq/output_payments.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+module Services
+  module Npq
+    class OutputPayment
+      include InitializeWithConfig
+      delegate :retention_points, :per_participant_price, to: :config
+
+      def call
+        retention_points.transform_values do |values|
+          {
+            per_participant: per_participant_output_payments,
```
Comment (blackrat) [inline lib/services/npq/output_payments.rb:12]: Singular for consistency again.

File: lib/services/npq/payment_calculation.rb [hunk 6]
```diff
@@ -10,7 +10,7 @@ def call
           input: config.to_h,
           output: {
             service_fees: ServiceFees.call(config),
-            variable_payments: VariablePayments.call(config),
+            output_payments: OutputPayment.call(config),
```
Comment (blackrat) [inline lib/services/npq/payment_calculation.rb:13]: Singular

File: spec/lib/services/ecf/payment_calculation_spec.rb [hunk 7]
```diff
@@ -21,9 +21,9 @@
     expect(result.dig(:output, :service_fees, :service_fee_per_participant)).to be_a(BigDecimal)
     expect(result.dig(:output, :service_fees, :service_fee_total)).to be_a(BigDecimal)
     expect(result.dig(:output, :service_fees, :service_fee_monthly)).to be_a(BigDecimal)
-    expect(result.dig(:output, :variable_payments, :per_participant)).to be_a(BigDecimal)
-    if result[:output][:variable_payments]
-      result.dig(:output, :variable_payments, :variable_payment_schedule).each do |_key, value|
+    expect(result.dig(:output, :output_payments, :per_participant)).to be_a(BigDecimal)
```
Comment (blackrat) [inline spec/lib/services/ecf/payment_calculation_spec.rb:24]: If we made the others singular, we need to do that here too.

File: spec/steps/ecf/payment_calculation_steps.rb [hunk 8]
```diff
@@ -14,8 +14,8 @@ module PaymentCalculationSteps
       {
         payment_type: values["Payment Type"],
         retained_participants: values["Retained Participants"].to_i,
-        expected_per_participant_variable_payment: CurrencyParser.currency_to_big_decimal(values["Expected Per-Participant Output Payment"]),
-        expected_variable_payment_subtotal: CurrencyParser.currency_to_big_decimal(values["Expected Output Payment Subtotal"]),
+        expected_per_participant_output_payments: CurrencyParser.currency_to_big_decimal(values["Expected Per-Participant Output Payment"]),
```
Comment (blackrat) [inline spec/steps/ecf/payment_calculation_steps.rb:17]: Less worried about the tests, but this is now inconsistent with its own heading

File: spec/steps/npq/payment_calculation_steps.rb [hunk 9]
```diff
@@ -70,16 +70,16 @@ def calculate
...
 
-  step "expected variable payments should be as above" do
+  step "expected output payments should be as above" do
     result = calculate
-    aggregate_failures "variable payments" do
+    aggregate_failures "output payments" do
       @retention_table.each do |retention_point, values|
         expect_with_context(
-          result.dig(:output, :variable_payments, retention_point, :per_participant), values[:expected_per_participant_variable_payment], "Payment for retention point '#{retention_point}'"
+          result.dig(:output, :output_payments, retention_point, :per_participant), values[:expected_per_participant_output_payments], "Payment for retention point '#{retention_point}'"
```
Comment (blackrat) [inline spec/steps/npq/payment_calculation_steps.rb:78]: singluar

**Review by blackrat — CHANGES_REQUESTED** [review]

**Review by blackrat — APPROVED** [review]

**Pull request merged**